src/wasm_udf/wasm.rs

```diff
@@ -11,22 +15,518 @@ use datafusion::{
+        let result: (Value, i32) = (output, (size + SIZE_BYTE_COUNT).try_into().unwrap());
+        Ok(result)
+    }
+
+    fn write_udf_input(&mut self, input: &Value) -> anyhow::Result<(i32, i32)> {
+        // serialize input using MessagePack
+        let mut udf_input_buf: Vec<u8> = vec![];
+        input
+            .serialize(&mut Serializer::new(&mut udf_input_buf))
+            .map_err(|err| {
+                anyhow::anyhow!(format!("Error serializing input {:?}", err))
+            })?;
+        // Total input size will be serialized messagepack bytes prepended by the
+        // size of the serialized input (one i32 == 4 bytes)
+        let size_len = std::mem::size_of::<i32>();
+        let udf_input_size: usize = udf_input_buf.len() + size_len;
+        // allocate WASM memory for input buffer
+        // TODO: handle case when allocation fails
+        let udf_input_ptr = self
+            .alloc
+            .call(&mut self.store, udf_input_size.try_into().unwrap())?;
+        let ptr: usize = udf_input_ptr.try_into().unwrap();
+        // write size of input buffer first
+        self.memory
+            .write(&mut self.store, ptr, &udf_input_buf.len().to_ne_bytes())
+            .map_err(|err| {
+                anyhow::anyhow!(format!("Error copying UDF input: {:?}", err))
+            })?;
+        // copy input buffer
+        self.memory
+            .write(&mut self.store, ptr + size_len, udf_input_buf.as_ref())
+            .map_err(|err| {
+                anyhow::anyhow!(format!("Error copying UDF input: {:?}", err))
+            })?;
+        // return the entire size of the output buffer (including i32 size prefix) so it can be passed to dealloc() later
+        Ok((udf_input_ptr, udf_input_size.try_into().unwrap()))
+    }
+
+    pub fn call(&mut self, input: Vec<Value>) -> anyhow::Result<Value> {
+        let args_array = Value::Array(input);
+        let (udf_input_ptr, input_size) = self.write_udf_input(&args_array)?;
+        // invoke UDF
+        let udf_output_ptr = self.udf.call(&mut self.store, udf_input_ptr)?;
+        let (output, output_size) = self.read_udf_output(udf_output_ptr)?;
+        // deallocate both input and output buffers
+        self.dealloc
+            .call(&mut self.store, (udf_input_ptr, input_size))?;
+        self.dealloc
+            .call(&mut self.store, (udf_output_ptr, output_size))?;
+        Ok(output)
+    }
+}
+
+fn get_arrow_value<T>(
+    args: &[ArrayRef],
+    row_ix: usize,
+    col_ix: usize,
+) -> Result<T::Native>
+where
+    T: ArrowPrimitiveType,
+{
+    match args
+        .get(col_ix)
+        .unwrap()
+        .as_any()
+        .downcast_ref::<PrimitiveArray<T>>()
+    {
+        Some(arr) => Ok(arr.value(row_ix)),
+        None => Err(DataFusionError::Internal(format!(
+            "Error casting column {:?} to array of primitive values",
+            col_ix
         ))),
     }
```

Comment:
I rewrote it a bit to avoid the match (mostly a stylistic choice, I didn't check if this compiles). Note that because of the `unwrap()` this function will panic if row/col_ix are out of bounds, so it's not completely "pure".

```suggestion
    let arr = args
        .get(col_ix)
        .unwrap()
        .as_any()
        .downcast_ref::<PrimitiveArray<T>>()
        .ok_or_else(|| DataFusionError::Internal(format!(
            "Error casting column {:?} to array of primitive values",
            col_ix
        )))?;
    Ok(arr.value(row_ix))
```